<a href="https://colab.research.google.com/github/AlinaZakharova1997/miramedix_practice/blob/main/%D0%9E%D1%86%D0%B5%D0%BD%D0%BA%D0%B0_%D1%82%D1%80%D0%B0%D0%BD%D1%81%D1%84%D0%BE%D1%80%D0%BC%D0%B5%D1%80%D0%B0_en_de_%D0%BC%D0%B0%D0%BB%D0%B5%D0%BD%D1%8C%D0%BA%D0%B0%D1%8F_%D0%B2%D1%8B%D0%B1%D0%BE%D1%80%D0%BA%D0%B0.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Install and import things we need

In [1]:
!git clone https://github.com/huggingface/transformers.git # установка и импорт библиотек (эту ячейку достаточно запустить один раз)
%cd transformers
!pip install . -q
!pip install sentencepiece -q
!pip install transformers==4.10

Cloning into 'transformers'...
remote: Enumerating objects: 99205, done.
remote: Total 99205 (delta 0), reused 0 (delta 0), pack-reused 99205
Receiving objects: 100% (99205/99205), 84.29 MiB | 24.46 MiB/s, done.
Resolving deltas: 100% (72001/72001), done.
/content/transformers
  DEPRECATION: A future pip version will change local packages to be built in-place without first copying to a temporary directory. We recommend you use --use-feature=in-tree-build to test your packages with this new behavior before it becomes the default.
   pip 21.3 will remove support for this functionality. You can find discussion regarding this at https://github.com/pypa/pip/issues/7555.
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
    Preparing wheel metadata ... done
     |████████████████████████████████| 596 kB 5.1 MB/s 
     |████████████████████████████████| 6.8 MB 43.5 MB/s 
     |████████████████████████████████| 67 kB 4.8 MB/s 
     |███████████████████████

In [ ]:
!pip install transformers==4.8

     |████████████████████████████████| 2.5 MB 12.9 MB/s 
  Attempting uninstall: huggingface-hub
    Found existing installation: huggingface-hub 0.4.0
    Uninstalling huggingface-hub-0.4.0:
      Successfully uninstalled huggingface-hub-0.4.0
  Attempting uninstall: transformers
    Found existing installation: transformers 4.10.0
    Uninstalling transformers-4.10.0:
      Successfully uninstalled transformers-4.10.0


In [2]:
from transformers import MarianMTModel, MarianTokenizer # (эту ячейку достаточно запустить один раз)
import torch

In [3]:
!pip install jsonlines
import jsonlines

In [4]:
import json

In [5]:
!pip install --upgrade !pip &&
!pip install unbabel-comet==1.0.1 --use-feature=2020-resolver

/bin/bash: -c: line 1: syntax error: unexpected end of file
     |████████████████████████████████| 56 kB 2.5 MB/s 
     |████████████████████████████████| 90 kB 5.3 MB/s 
     |████████████████████████████████| 101 kB 10.1 MB/s 
     |████████████████████████████████| 15.7 MB 49.6 MB/s 
     |████████████████████████████████| 329 kB 37.3 MB/s 
     |████████████████████████████████| 808 kB 33.4 MB/s 
     |████████████████████████████████| 38.1 MB 3.6 MB/s 
     |████████████████████████████████| 9.5 MB 35.6 MB/s 
     |████████████████████████████████| 636 kB 37.8 MB/s 
     |████████████████████████████████| 133 kB 35.0 MB/s 
     |████████████████████████████████| 829 kB 31.8 MB/s 
     |████████████████████████████████| 1.1 MB 43.1 MB/s 
     |████████████████████████████████| 271 kB 53.8 MB/s 
     |████████████████████████████████| 144 kB 51.1 MB/s 
     |████████████████████████████████| 94 kB 2.6 MB/s 
  Created wheel for future: filename=future-0.18.2-py3-none-any.whl size=49

In [6]:
import comet
from comet import download_model, load_from_checkpoint

In [7]:
# COMET_MODEL = "wmt20-comet-da"
COMET_MODEL = "wmt21-cometinho-da"
model_path = download_model(COMET_MODEL)
comet_model = load_from_checkpoint(model_path)

wmt21-cometinho-da.tar.gz: 344MB [00:10, 32.3MB/s]                           
Extracting /root/.cache/torch/unbabel_comet/wmt21-cometinho-da.tar.gz
Extracted /root/.cache/torch/unbabel_comet/wmt21-cometinho-da.tar.gz


Downloading:   0%|          | 0.00/5.07M [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/239 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/392 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/9.08M [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/689 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/427M [00:00<?, ?B/s]

Encoder model frozen.


In [8]:
!pip install datasets

     |████████████████████████████████| 311 kB 5.1 MB/s 
     |████████████████████████████████| 243 kB 49.4 MB/s 


Metrics we will use

In [9]:
!pip install rouge_score

In [10]:
!pip install sacrebleu

In [11]:
!pip install -U nltk

     |████████████████████████████████| 1.5 MB 5.5 MB/s 
     |████████████████████████████████| 748 kB 46.6 MB/s 
  Attempting uninstall: regex
    Found existing installation: regex 2019.12.20
    Uninstalling regex-2019.12.20:
      Successfully uninstalled regex-2019.12.20
  Attempting uninstall: nltk
    Found existing installation: nltk 3.2.5
    Uninstalling nltk-3.2.5:
      Successfully uninstalled nltk-3.2.5


In [12]:
import nltk
import numpy
import random

In [13]:
from datasets import load_metric, list_metrics
metrics_list = list_metrics()
#print(metrics_list)

bleu_metric = load_metric('bleu')
rouge_metric = load_metric('rouge')
sacrebleu_metric = load_metric('sacrebleu')
meteor_metric = load_metric('meteor')

Downloading:   0%|          | 0.00/2.48k [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/1.55k [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/2.16k [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/2.36k [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/2.19k [00:00<?, ?B/s]

[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Unzipping corpora/wordnet.zip.
[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.
[nltk_data] Downloading package omw-1.4 to /root/nltk_data...
[nltk_data]   Unzipping corpora/omw-1.4.zip.


Model we will test

In [14]:
DEVICE = 'cuda' if torch.cuda.is_available() else 'cpu'

def load_model(model_name):
  tokenizer = MarianTokenizer.from_pretrained(model_name)
  model = MarianMTModel.from_pretrained(model_name).to(DEVICE)
  return tokenizer, model

Тут будут вспомогательные функции для ускорения работы

Эти функции нужны для преобразования результатов для вычисления метрики bleu

In [15]:
# We need this to prepare outputs for bleu metric

def ref_to_bleu(beam_reference):
  beam_ref = []
  for item in beam_reference:
    text = item[0]
    tokenized = tokenizer.tokenize(text)
    beam_ref.append([tokenized])
  return beam_ref

def hyp_to_bleu(beam_hypothesis):
  beam_hyp = []
  for text in beam_hypothesis:
    tokenized = tokenizer.tokenize(text)
    beam_hyp.append(tokenized)
  return beam_hyp

Эти функции нужны для преобразования данных для вычисление метрики comet

In [16]:
def comet_data_maker(source, hypothesis, reference):
  comet_reference = []
  for d in reference:
    comet_reference.append(d[0])
  comet_data = [{"src": src, "mt": mt, "ref": ref} for src, mt, ref in
                  zip(source, hypothesis, comet_reference)]
  return comet_data

Эта функция будет вычислять метрики, которые мы выбрали для оценки модели
Мы взяли BLEU, ROUGE, SACREBLEU, METEOR, COMET 

In [17]:
BLEU = 'bleu'
ROUGE = 'rouge'
SACREBLEU = 'sacrebleu'
METEOR = 'meteor'
COMET = 'comet'

In [18]:
def compute_metric(metric_name, references, predictions):
    return load_metric(metric_name).compute(predictions=predictions,
                                            references=references)

Эта функция для того, чтобы посчитать все нужные метрики и записать их в отдельный json file

In [19]:
def get_results(filename, source, hypothesis, reference):

  path = '/content/drive/MyDrive/'+filename
  comet_data = comet_data_maker(source, hypothesis, reference)
  bleu_references = ref_to_bleu(reference)
  bleu_predictions = hyp_to_bleu(hypothesis)

  results = {BLEU: compute_metric(BLEU, bleu_references, bleu_predictions),
               ROUGE: compute_metric(ROUGE, reference, hypothesis),
               SACREBLEU: compute_metric(SACREBLEU, reference,
                                         hypothesis),
               METEOR: compute_metric(METEOR, reference, hypothesis),
               COMET: comet_model.predict(comet_data, gpus=0),
               }
  with open(path, 'w', encoding='utf8') as f:
    json.dump(results, f)
  return results

Результат работы модели английский-немецкий.




Здесь импортировали дообученную модель с гугл диска
Взяли 100 случайных пар англ-нем тестовых текстов, чтобы на них посмотреть как справится модель 

In [20]:
from google.colab import drive
drive.mount('/content/drive') # (эту ячейку достаточно запустить один раз)

Mounted at /content/drive


In [21]:
model_name = '/content/drive/MyDrive/model_en_de_15ep_output/'
tokenizer, model = load_model(model_name)

In [22]:
with jsonlines.open('/content/drive/MyDrive/test_data.json', 'r') as reader:
    target = []
    src = []
    values = []
    for obj in reader:
      values.append(obj)
    for i in range(100):
      random_pair = random.choice(values)
      target.append(random_pair['translation']['de'])
      src.append(random_pair['translation']['en'])

# train_data.train_data.to(torch.device("cuda:0"))

In [22]:
target[0]


'Diese Versorgungskette beginnt mit einer gemeindenahen Initiative, die durch die nichtstaatliche Organisation (NGO) Carusel durchgeführt wird. Deren Mitarbeiter sprechen mit Menschen, die auf den Straßen Bukarests leben und arbeiten, beraten diese, wo und wann sie Gesundheitsleistungen in Anspruch nehmen können (darunter auch Impfungen), und verweisen sie an soziale Einrichtungen und Zentren, um sich für eine COVID-19-Impfung registrieren zu lassen.'

In [23]:
src[0]

'This chain of care begins with community outreach undertaken by the nongovernmental organization (NGO) Carusel. Its staff speak to people living and working on Bucharest’s streets, advising them on when and where they can receive health services, including vaccinations, and referring them to social-care centres to register for the COVID-19 vaccination.'

Посчитали среднюю длину текстов в таргете

In [27]:
len_text = []
# print(target[0])
for text in target:

  l = len(text)
  len_text.append(l)
print(numpy.mean(len_text))

477.93


Greedy search

Это просто самый обычный greedy search, тут из параметров только максимальная длина, которая собств максимальная 512, чтобы не точно резать предложения.

In [23]:
source = []
hypothesis = []
reference = []
compare = {}

In [24]:
for i in range(len(target)):

  target_item = target[i] # reference
  #print(target_item, 'TARGET')
  src_item = src[i] # source
  #print(src_item, 'src_item')
  inputs = tokenizer.encode(src_item, return_tensors="pt", max_length=512, truncation=True)
  #print(inputs, 'inputs')
  greedy_outputs = model.generate(inputs.cuda()) 
  result = tokenizer.decode(greedy_outputs[0], skip_special_tokens=True) # hypothesis
  compare[target_item] = result
  #print(result, 'RESULT')

  # collecting data for metrics count
  source.append(src_item)
  hypothesis.append(result)
  reference.append([target_item])
  

In [32]:
source

['The first in a series of simulated exercises will show the critical role rapid-response mobile laboratories, known as Public Health RRMLs, play in fighting public health emergencies by enabling onsite testing and analysis during outbreaks.',
 'Ten days after the implementation of broad public health and social measures, cases began to decline in Germany. Mortality rates and the median age of confirmed cases in Germany are lower than the average elsewhere. This is linked to a range of factors – including population demography and widespread testing.',
 'Mental health challenges remain deeply stigmatized in many communities. Community sensitization to the spectrum of mental health challenges will be important as the true scale of the crisis becomes clearer. Specialized care and support services need to be sustained for those in the grip of more severe mental health challenges.',
 'It is therefore vital to take a long-term view and to use that perspective to strengthen health policies a

In [26]:
reference

[['Die erste in einer Reihe von Simulationsübungen soll die wichtige Rolle verdeutlichen, die mobile Labore für den schnellen Feldeinsatz, sogenannte Rapid-Response Mobile Laboratories (RRML), bei der Eindämmung gesundheitlicher Notlagen spielen, indem sie während Krankheitsausbrüchen die Durchführung von Tests und Analysen direkt vor Ort ermöglichen.'],
 ['Doch auch innerhalb von Ländern sind die Fälle nicht gleichmäßig verteilt. So tragen in Italien die Lombardei und in Spanien Madrid jeweils die Hauptlast der Pandemie. Glücklicherweise nimmt die Zahl der täglichen Todesfälle in beiden Regionen inzwischen ab.'],
 ['Psychische Gesundheitsprobleme sind in unseren Gesellschaften oftmals noch mit einem starken Stigma verbunden. Eine Sensibilisierung der Gesellschaft für die große Bandbreite an psychischen Gesundheitsproblemen wird nun, da sich das volle Ausmaß der Krise allmählich offenbart, immer wichtiger. Für Personen, die unter schwereren psychischen Problemen leiden, müssen speziali

In [25]:
hypothesis

['Die erste in einer Reihe von Simulationsübungen soll die wichtige Rolle verdeutlichen, die mobile Labore für den schnellen Feldeinsatz, sogenannte Rapid-Response Mobile Laboratories (RRML), bei der Eindämmung gesundheitlicher Notlagen spielen, indem sie während Krankheitsausbrüchen die Durchführung von Tests und Analysen direkt vor Ort ermöglichen.',
 'Zehn Tage nach Einführung umfangreicher gesundheitlicher und sozialer Maßnahmen sank die Zahl der Fälle in Deutschland. Die Sterblichkeitsraten und das Durchschnittsalter der bestätigten Fälle sind in Deutschland niedriger als im Durchschnitt andererorts. Dies geht mit einer Reihe von Faktoren einher, darunter die demografische Demografie und die flächendeckende Testung.',
 'Psychische Herausforderungen bleiben in vielen Gemeinschaften zutiefst stigmatisiert. Eine Sensibilisierung der Gemeinschaft für das Spektrum der psychischen Herausforderungen wird wichtig sein, da die wahre Größenordnung der Krise klarer wird. Spezialisierte Verso

Save greedy data

In [56]:
get_results('greedy_search_results.json', source, hypothesis, reference)

[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package omw-1.4 to /root/nltk_data...
[nltk_data]   Package omw-1.4 is already up-to-date!
GPU available: True, used: False
TPU available: False, using: 0 TPU cores
/usr/local/lib/python3.7/dist-packages/pytorch_lightning/utilities/distributed.py:69: UserWarning: GPU available but not used. Set the gpus flag in your trainer `Trainer(gpus=1)` or script `--gpus=1`.
  warnings.warn(*args, **kwargs)
Predicting: 100%|██████████| 13/13 [04:52<00:00, 22.51s/it]


{'bleu': {'bleu': 0.5032662609906366,
  'brevity_penalty': 0.841612003138467,
  'length_ratio': 0.852924893891862,
  'precisions': [0.7794605509880282,
   0.6218218799941886,
   0.5433923606029563,
   0.48547840188707064],
  'reference_length': 16257,
  'translation_length': 13866},
 'comet': ([0.6994382739067078,
   0.3296681344509125,
   0.7267894148826599,
   0.6021633744239807,
   0.5547754764556885,
   -0.7143052220344543,
   0.06844544410705566,
   0.8460600972175598,
   0.7420597672462463,
   0.7546868920326233,
   0.6464825868606567,
   0.8002039790153503,
   0.6396624445915222,
   -0.979827344417572,
   0.5862824320793152,
   0.07118101418018341,
   0.8054172992706299,
   0.6181743144989014,
   -0.8671560287475586,
   0.829494297504425,
   0.7002013921737671,
   0.7564400434494019,
   0.5486039519309998,
   0.872108519077301,
   0.7680116295814514,
   0.8009846210479736,
   0.7001193165779114,
   0.7379111647605896,
   0.6013379693031311,
   0.7879781126976013,
   0.4277176856

bleu : 0.5032662609906366

comet: 0.5405535845458508

meteor: 0.559520092773177

sacrebleu: 41.274459158341


In [36]:
results, f = get_results('greedy_search_results', source, hypothesis, reference)

[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package omw-1.4 to /root/nltk_data...
[nltk_data]   Package omw-1.4 is already up-to-date!
GPU available: True, used: False
TPU available: False, using: 0 TPU cores
/usr/local/lib/python3.7/dist-packages/pytorch_lightning/utilities/distributed.py:69: UserWarning: GPU available but not used. Set the gpus flag in your trainer `Trainer(gpus=1)` or script `--gpus=1`.
  warnings.warn(*args, **kwargs)
Predicting: 100%|██████████| 13/13 [04:45<00:00, 21.94s/it]


Beam

попробуем посмотреть, станет ли еще лучше. Добавим раннюю остановку, чтобы чтобы генерация завершалась, когда все гипотезы луча достигают токена EOS, число бимов, число возвращаемых предложений, поставим больше 1

In [27]:
beam_source = []
beam_hypothesis = []
beam_reference = []
beam_compare = {}

In [28]:
for i in range(len(target)):

  target_item = target[i] # reference
  #print(target_item, 'TARGET')
  src_item = src[i] # source
  #print(src_item, 'src_item')
  inputs = tokenizer.encode(src_item, return_tensors="pt", max_length=512, truncation=True)
  #print(inputs, 'inputs')
  
  beam_outputs = model.generate(
    inputs.cuda(), 
    num_beams=5, 
    num_return_sequences=5,
    early_stopping=True,
)
  result = tokenizer.decode(beam_outputs[0], skip_special_tokens=True) # hypothesis
  beam_compare[target_item] = result
  #print(result, 'RESULT')

  # collecting data for metrics count
  beam_source.append(src_item)
  beam_hypothesis.append(result)
  beam_reference.append([target_item])
  

In [29]:
beam_reference

[['Die erste in einer Reihe von Simulationsübungen soll die wichtige Rolle verdeutlichen, die mobile Labore für den schnellen Feldeinsatz, sogenannte Rapid-Response Mobile Laboratories (RRML), bei der Eindämmung gesundheitlicher Notlagen spielen, indem sie während Krankheitsausbrüchen die Durchführung von Tests und Analysen direkt vor Ort ermöglichen.'],
 ['Doch auch innerhalb von Ländern sind die Fälle nicht gleichmäßig verteilt. So tragen in Italien die Lombardei und in Spanien Madrid jeweils die Hauptlast der Pandemie. Glücklicherweise nimmt die Zahl der täglichen Todesfälle in beiden Regionen inzwischen ab.'],
 ['Psychische Gesundheitsprobleme sind in unseren Gesellschaften oftmals noch mit einem starken Stigma verbunden. Eine Sensibilisierung der Gesellschaft für die große Bandbreite an psychischen Gesundheitsproblemen wird nun, da sich das volle Ausmaß der Krise allmählich offenbart, immer wichtiger. Für Personen, die unter schwereren psychischen Problemen leiden, müssen speziali

In [30]:
beam_hypothesis

['Die erste in einer Reihe von Simulationsübungen soll die entscheidende Rolle verdeutlichen, die mobile Labore für den schnellen Feldeinsatz, sogenannte Rapid-Response Mobile Laboratories (RRML), bei der Eindämmung gesundheitlicher Notlagen spielen, indem sie während Krankheitsausbrüchen die Durchführung von Tests und Analysen direkt vor Ort ermöglichen',
 'Zehn Tage nach Einführung breiter gesundheitlicher und sozialer Maßnahmen sank die Zahl der Fälle in Deutschland. Die Sterblichkeitsraten und das Durchschnittsalter der bestätigten Fälle sind in Deutschland niedriger als im Durchschnitt andererorts. Dies geht mit einer Reihe von Faktoren einher, darunter die demografische Demografie und die flächendeckende Testung.',
 'Psychische Herausforderungen bleiben in vielen Gemeinschaften zutiefst stigmatisiert. Eine Sensibilisierung der Gemeinschaft für das Spektrum der psychischen Herausforderungen wird wichtig sein, da die wahre Größenordnung der Krise klarer wird. Spezialisierte Versorg

In [31]:
beam_source

['The first in a series of simulated exercises will show the critical role rapid-response mobile laboratories, known as Public Health RRMLs, play in fighting public health emergencies by enabling onsite testing and analysis during outbreaks.',
 'Ten days after the implementation of broad public health and social measures, cases began to decline in Germany. Mortality rates and the median age of confirmed cases in Germany are lower than the average elsewhere. This is linked to a range of factors – including population demography and widespread testing.',
 'Mental health challenges remain deeply stigmatized in many communities. Community sensitization to the spectrum of mental health challenges will be important as the true scale of the crisis becomes clearer. Specialized care and support services need to be sustained for those in the grip of more severe mental health challenges.',
 'It is therefore vital to take a long-term view and to use that perspective to strengthen health policies a

In [60]:
get_results('beam_results.json', beam_source, beam_hypothesis, beam_reference)

[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package omw-1.4 to /root/nltk_data...
[nltk_data]   Package omw-1.4 is already up-to-date!
GPU available: True, used: False
TPU available: False, using: 0 TPU cores
/usr/local/lib/python3.7/dist-packages/pytorch_lightning/utilities/distributed.py:69: UserWarning: GPU available but not used. Set the gpus flag in your trainer `Trainer(gpus=1)` or script `--gpus=1`.
  warnings.warn(*args, **kwargs)
Predicting: 100%|██████████| 13/13 [04:51<00:00, 22.39s/it]


{'bleu': {'bleu': 0.5011909810190921,
  'brevity_penalty': 0.8409713704736994,
  'length_ratio': 0.8523712862151689,
  'precisions': [0.7791008154723245,
   0.6197572145089772,
   0.5407483341875962,
   0.48314523862211406],
  'reference_length': 16257,
  'translation_length': 13857},
 'comet': ([0.6994889378547668,
   0.3296681344509125,
   0.7267945408821106,
   0.604601263999939,
   0.5547928810119629,
   -0.7143028378486633,
   0.06879439204931259,
   0.8462262749671936,
   0.7420628070831299,
   0.7546868920326233,
   0.6464825868606567,
   0.8001907467842102,
   0.6398218274116516,
   -0.979827344417572,
   0.6495024561882019,
   0.11894681304693222,
   0.8054168820381165,
   0.6181640028953552,
   -0.8672868609428406,
   0.8294969201087952,
   0.6983943581581116,
   0.7564507126808167,
   0.5235781073570251,
   0.872108519077301,
   0.7680116295814514,
   0.8009821772575378,
   0.7001193165779114,
   0.7378948330879211,
   0.6013379693031311,
   0.7879781126976013,
   0.42812573

Тут значения метрик несильно но уменьшилось, не самый лучший результат.

More experiments

Тут все тоже самое, но еще добавляем параметр no_repeat_ngram_size=2, который исключает повторения биграмм, так по идее тексты должны стать еще лучше и метрики вырастут

In [33]:
beam_source = []
beam_hypothesis = []
beam_reference = []
beam_compare = {}

In [35]:
for i in range(len(target)):

  target_item = target[i] # reference
  #print(target_item, 'target_item')
  #print(target_item, 'TARGET')
  src_item = src[i] # source
  #print(src_item, 'src_item')
  #print(src_item, 'src_item')
  inputs = tokenizer.encode(src_item, return_tensors="pt", max_length=512, truncation=True)
  #print(inputs, 'inputs')
  #print(inputs, 'inputs')
  beam_output = model.generate(
    inputs.cuda(),  
    num_beams=5, 
    no_repeat_ngram_size=2, 
    num_return_sequences=5,
    early_stopping=True 
)
  result = tokenizer.decode(beam_output[0], skip_special_tokens=True) # hypothesis
  beam_compare[target_item] = result
  #print(result, 'RESULT')

  # collecting data for metrics count
  beam_source.append(src_item)
  beam_hypothesis.append(result)
  beam_reference.append([target_item])
  

In [36]:
beam_source

['The first in a series of simulated exercises will show the critical role rapid-response mobile laboratories, known as Public Health RRMLs, play in fighting public health emergencies by enabling onsite testing and analysis during outbreaks.',
 'Ten days after the implementation of broad public health and social measures, cases began to decline in Germany. Mortality rates and the median age of confirmed cases in Germany are lower than the average elsewhere. This is linked to a range of factors – including population demography and widespread testing.',
 'Mental health challenges remain deeply stigmatized in many communities. Community sensitization to the spectrum of mental health challenges will be important as the true scale of the crisis becomes clearer. Specialized care and support services need to be sustained for those in the grip of more severe mental health challenges.',
 'It is therefore vital to take a long-term view and to use that perspective to strengthen health policies a

In [37]:
beam_hypothesis

['Die erste in einer Reihe von Simulationsübungen soll die entscheidende Rolle verdeutlichen, die mobile Labore für den schnellen Feldeinsatz, sogenannte Rapid-Response Mobile Laboratories (RRML), bei der Eindämmung gesundheitlicher Notlagen spielen, indem sie während Krankheitsausbrüchen die Durchführung von Tests und Analysen direkt vor Ort ermöglichen',
 'Zehn Tage nach Einführung breiter Maßnahmen zum Schutz der öffentlichen Gesundheit und sozialer Maßnahmen sank die Zahl der Fälle in Deutschland. Die Sterblichkeitsraten und das Durchschnittsalter der bestätigten Fälle sind in anderen Bundesländern niedriger als im Durchschnitt. Dies geht mit einer Reihe von Faktoren einher, darunter die demografische Demografie und die flächendeckende Testung.',
 'Die Herausforderungen für die psychische Gesundheit bleiben in vielen Gemeinschaften zutiefst stigmatisiert. Die Sensibilisierung der Gemeinschaft für das Spektrum der psychischen Gesundheitsprobleme wird von Bedeutung sein, da die wahre

In [38]:
beam_reference

[['Die erste in einer Reihe von Simulationsübungen soll die wichtige Rolle verdeutlichen, die mobile Labore für den schnellen Feldeinsatz, sogenannte Rapid-Response Mobile Laboratories (RRML), bei der Eindämmung gesundheitlicher Notlagen spielen, indem sie während Krankheitsausbrüchen die Durchführung von Tests und Analysen direkt vor Ort ermöglichen.'],
 ['Doch auch innerhalb von Ländern sind die Fälle nicht gleichmäßig verteilt. So tragen in Italien die Lombardei und in Spanien Madrid jeweils die Hauptlast der Pandemie. Glücklicherweise nimmt die Zahl der täglichen Todesfälle in beiden Regionen inzwischen ab.'],
 ['Psychische Gesundheitsprobleme sind in unseren Gesellschaften oftmals noch mit einem starken Stigma verbunden. Eine Sensibilisierung der Gesellschaft für die große Bandbreite an psychischen Gesundheitsproblemen wird nun, da sich das volle Ausmaß der Krise allmählich offenbart, immer wichtiger. Für Personen, die unter schwereren psychischen Problemen leiden, müssen speziali

In [27]:
get_results('beam_no_repeat_ngram_2_results.json', beam_source,beam_hypothesis,beam_reference)

[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package omw-1.4 to /root/nltk_data...
[nltk_data]   Package omw-1.4 is already up-to-date!
GPU available: False, used: False
TPU available: False, using: 0 TPU cores
Predicting: 100%|██████████| 13/13 [05:54<00:00, 27.30s/it]


{'bleu': {'bleu': 0.5325269662090122,
  'brevity_penalty': 0.9110339878012991,
  'length_ratio': 0.9147665580890336,
  'precisions': [0.7670623145400594,
   0.6100149476831092,
   0.5295180722891566,
   0.47116843702579664],
  'reference_length': 14736,
  'translation_length': 13480},
 'comet': ([0.6234658360481262,
   0.6050043106079102,
   1.2659318447113037,
   0.6087995171546936,
   0.6293062567710876,
   0.8900836706161499,
   0.7652810215950012,
   0.6463749408721924,
   0.7512192130088806,
   0.6258825659751892,
   0.6176605820655823,
   0.692544162273407,
   0.8671794533729553,
   0.7422592043876648,
   0.8570452332496643,
   0.07222238928079605,
   -1.8186508417129517,
   0.669430673122406,
   0.6768596768379211,
   0.4765022397041321,
   0.773017168045044,
   0.4011501669883728,
   0.719650387763977,
   0.538377046585083,
   0.18651729822158813,
   0.3077225089073181,
   0.577340304851532,
   0.6797649264335632,
   1.0059137344360352,
   0.20618024468421936,
   0.801096141338

bleu: 0.5325269662090122

sacrebleu: 43.036563804150326

meteor: 0.5788023114617102

comet: 0.5354018643312156

получше стало, чем предыдущие два, но еще посмотрим, может еще увеличим

K sample

посмотрим, как работает семпл, поставим пока k=0

In [39]:
k_source = []
k_hypothesis = []
k_reference = []
k_compare = {}

In [41]:
for i in range(len(target)):

  target_item = target[i] # reference
  #print(target_item, 'TARGET')
  src_item = src[i] # source
  #print(src_item, 'src_item')
  inputs = tokenizer.encode(src_item, return_tensors="pt", max_length=512, truncation=True)
  #print(inputs, 'inputs')
  
  sample_output = model.generate(
    inputs.cuda(), 
    do_sample=True, 
    max_length=512, 
    top_k=0,
    early_stopping=True
)
  result = tokenizer.decode(sample_output[0], skip_special_tokens=True) # hypothesis
  k_compare[target_item] = result
  #print(result, 'RESULT')

  # collecting data for metrics count
  k_source.append(src_item)
  k_hypothesis.append(result)
  k_reference.append([target_item])
  

/usr/local/lib/python3.7/dist-packages/transformers/generation_utils.py:2138: UserWarning: __floordiv__ is deprecated, and its behavior will change in a future version of pytorch. It currently rounds toward 0 (like the 'trunc' function NOT 'floor'). This results in incorrect rounding for negative values. To keep the current behavior, use torch.div(a, b, rounding_mode='trunc'), or for actual floor division, use torch.div(a, b, rounding_mode='floor').
  next_indices = next_tokens // vocab_size


In [42]:
k_source

['The first in a series of simulated exercises will show the critical role rapid-response mobile laboratories, known as Public Health RRMLs, play in fighting public health emergencies by enabling onsite testing and analysis during outbreaks.',
 'Ten days after the implementation of broad public health and social measures, cases began to decline in Germany. Mortality rates and the median age of confirmed cases in Germany are lower than the average elsewhere. This is linked to a range of factors – including population demography and widespread testing.',
 'Mental health challenges remain deeply stigmatized in many communities. Community sensitization to the spectrum of mental health challenges will be important as the true scale of the crisis becomes clearer. Specialized care and support services need to be sustained for those in the grip of more severe mental health challenges.',
 'It is therefore vital to take a long-term view and to use that perspective to strengthen health policies a

In [43]:
k_hypothesis

['Die erste in einer Reihe von Simulationsübungen soll die wichtige Rolle verdeutlichen, die mobile Labore für den schnellen Feldeinsatz, sogenannte Rapid-Response Mobile Laboratories (RRML), bei der Eindämmung gesundheitlicher Notlagen spielen, indem sie während Krankheitsausbrüchen die Durchführung von Tests und Analysen direkt vor Ort ermöglichen.',
 'Zehn Tage nach Einführung breiter Maßnahmen zum Schutz der öffentlichen Gesundheit und sozialer Maßnahmen sank die Zahl der Fälle in Deutschland. Die Sterblichkeitsraten und das Durchschnittsalter der bestätigten Fälle sind in Deutschland geringer als im Durchschnitt andererorts. Dies ist mit einer Reihe von Faktoren verknüpft, darunter die demografische Demografie und die flächendeckende Testung.',
 'Psychische Herausforderungen bleiben in vielen Gemeinschaften zutiefst stigmatisiert. Die Sensibilisierung der Gemeinschaft für das Spektrum der psychischen Herausforderungen wird von Bedeutung sein, da die wahre Größenordnung der Krise k

In [44]:
k_reference

[['Die erste in einer Reihe von Simulationsübungen soll die wichtige Rolle verdeutlichen, die mobile Labore für den schnellen Feldeinsatz, sogenannte Rapid-Response Mobile Laboratories (RRML), bei der Eindämmung gesundheitlicher Notlagen spielen, indem sie während Krankheitsausbrüchen die Durchführung von Tests und Analysen direkt vor Ort ermöglichen.'],
 ['Doch auch innerhalb von Ländern sind die Fälle nicht gleichmäßig verteilt. So tragen in Italien die Lombardei und in Spanien Madrid jeweils die Hauptlast der Pandemie. Glücklicherweise nimmt die Zahl der täglichen Todesfälle in beiden Regionen inzwischen ab.'],
 ['Psychische Gesundheitsprobleme sind in unseren Gesellschaften oftmals noch mit einem starken Stigma verbunden. Eine Sensibilisierung der Gesellschaft für die große Bandbreite an psychischen Gesundheitsproblemen wird nun, da sich das volle Ausmaß der Krise allmählich offenbart, immer wichtiger. Für Personen, die unter schwereren psychischen Problemen leiden, müssen speziali

In [27]:
get_results('k_samples_results.json', k_source,k_hypothesis,k_reference)

[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package omw-1.4 to /root/nltk_data...
[nltk_data]   Package omw-1.4 is already up-to-date!
GPU available: False, used: False
TPU available: False, using: 0 TPU cores
Predicting: 100%|██████████| 13/13 [06:11<00:00, 28.60s/it]


{'bleu': {'bleu': 0.4810456248925994,
  'brevity_penalty': 0.8413523664437557,
  'length_ratio': 0.8527004909983633,
  'precisions': [0.7488675623800384,
   0.6006963249516442,
   0.5178947368421053,
   0.4587033398821218],
  'reference_length': 15275,
  'translation_length': 13025},
 'comet': ([0.20457032322883606,
   0.2860533893108368,
   -1.0478808879852295,
   0.6413853168487549,
   0.8147163987159729,
   0.7967235445976257,
   0.5662445425987244,
   0.5552389025688171,
   0.7876363396644592,
   0.6116360425949097,
   0.4496639370918274,
   0.8590372204780579,
   0.705149233341217,
   0.7203416228294373,
   0.715869128704071,
   0.22255414724349976,
   0.8985593914985657,
   0.6633207201957703,
   0.7275550365447998,
   -1.4832909107208252,
   0.6371985673904419,
   0.43738701939582825,
   0.9799920916557312,
   0.7577546238899231,
   0.6615412831306458,
   0.7125268578529358,
   0.613391101360321,
   0.5805377960205078,
   0.8561211228370667,
   0.5886645317077637,
   0.567250847

bleu: 0.4810456248925994

sacrebleu: 38.90251803660552

meteor: 0.5653108776532656

comet: 0.5670439285039902

Попроседало, докрутим еще, пусть k=50, num_return_sequences=3

In [45]:
k_source = []
k_hypothesis = []
k_reference = []
k_compare = {}

In [47]:
for i in range(len(target)):

  target_item = target[i] # reference
  #print(target_item, 'TARGET')
  src_item = src[i] # source
  #print(src_item, 'src_item')
  inputs = tokenizer.encode(src_item, return_tensors="pt", max_length=512, truncation=True)
  #print(inputs, 'inputs')
  
  sample_output = model.generate(
    inputs.cuda(), 
    do_sample=True, 
    max_length=512, 
    top_k=50,
    num_return_sequences=3,
    early_stopping=True
)
  result = tokenizer.decode(sample_output[0], skip_special_tokens=True) # hypothesis
  k_compare[target_item] = result
  #print(result, 'RESULT')

  # collecting data for metrics count
  k_source.append(src_item)
  k_hypothesis.append(result)
  k_reference.append([target_item])

/usr/local/lib/python3.7/dist-packages/transformers/generation_utils.py:2138: UserWarning: __floordiv__ is deprecated, and its behavior will change in a future version of pytorch. It currently rounds toward 0 (like the 'trunc' function NOT 'floor'). This results in incorrect rounding for negative values. To keep the current behavior, use torch.div(a, b, rounding_mode='trunc'), or for actual floor division, use torch.div(a, b, rounding_mode='floor').
  next_indices = next_tokens // vocab_size


In [48]:
k_source

['The first in a series of simulated exercises will show the critical role rapid-response mobile laboratories, known as Public Health RRMLs, play in fighting public health emergencies by enabling onsite testing and analysis during outbreaks.',
 'Ten days after the implementation of broad public health and social measures, cases began to decline in Germany. Mortality rates and the median age of confirmed cases in Germany are lower than the average elsewhere. This is linked to a range of factors – including population demography and widespread testing.',
 'Mental health challenges remain deeply stigmatized in many communities. Community sensitization to the spectrum of mental health challenges will be important as the true scale of the crisis becomes clearer. Specialized care and support services need to be sustained for those in the grip of more severe mental health challenges.',
 'It is therefore vital to take a long-term view and to use that perspective to strengthen health policies a

In [49]:
k_hypothesis

['Die erste in einer Reihe von Simulationsübungen soll die wichtige Rolle verdeutlichen, die mobile Labore für den schnellen Feldeinsatz, sogenannte Rapid-Response Mobile Laboratories (RRML), bei der Eindämmung gesundheitlicher Notlagen spielen, indem sie während Krankheitsausbrüchen die Durchführung von Tests und Analysen direkt vor Ort ermöglichen.',
 'Zehn Tage nach Einführung breiter gesundheitlicher und sozialer Maßnahmen sank die Zahl der Fälle in Deutschland. Die Sterblichkeitsraten und das Durchschnittsalter der bestätigten Fälle sind in Deutschland niedriger als im Durchschnitt andererorts. Dies ist mit einer Reihe von Faktoren verbunden, darunter die demografische Demografie und die flächendeckende Testung.',
 'Psychische Herausforderungen bleiben in vielen Gemeinschaften zutiefst stigmatisiert. Eine Sensibilisierung der Gemeinschaft für das Spektrum der psychischen Herausforderungen wird wichtig sein, da das ganze Ausmaß der Krise klarer wird. Spezialisierte Versorgungsmaßna

In [50]:
k_reference

[['Die erste in einer Reihe von Simulationsübungen soll die wichtige Rolle verdeutlichen, die mobile Labore für den schnellen Feldeinsatz, sogenannte Rapid-Response Mobile Laboratories (RRML), bei der Eindämmung gesundheitlicher Notlagen spielen, indem sie während Krankheitsausbrüchen die Durchführung von Tests und Analysen direkt vor Ort ermöglichen.'],
 ['Doch auch innerhalb von Ländern sind die Fälle nicht gleichmäßig verteilt. So tragen in Italien die Lombardei und in Spanien Madrid jeweils die Hauptlast der Pandemie. Glücklicherweise nimmt die Zahl der täglichen Todesfälle in beiden Regionen inzwischen ab.'],
 ['Psychische Gesundheitsprobleme sind in unseren Gesellschaften oftmals noch mit einem starken Stigma verbunden. Eine Sensibilisierung der Gesellschaft für die große Bandbreite an psychischen Gesundheitsproblemen wird nun, da sich das volle Ausmaß der Krise allmählich offenbart, immer wichtiger. Für Personen, die unter schwereren psychischen Problemen leiden, müssen speziali

In [31]:
get_results('k_50_samples_results.json', k_source,k_hypothesis,k_reference)

[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package omw-1.4 to /root/nltk_data...
[nltk_data]   Package omw-1.4 is already up-to-date!
GPU available: False, used: False
TPU available: False, using: 0 TPU cores
Predicting: 100%|██████████| 13/13 [06:03<00:00, 27.99s/it]


{'bleu': {'bleu': 0.4807550435459599,
  'brevity_penalty': 0.8390016900926912,
  'length_ratio': 0.8506710310965631,
  'precisions': [0.7511928582422657,
   0.6019078641228478,
   0.5191496013756448,
   0.4592720970537262],
  'reference_length': 15275,
  'translation_length': 12994},
 'comet': ([0.20457032322883606,
   0.30417755246162415,
   -1.2503044605255127,
   0.6413030028343201,
   0.8086679577827454,
   0.8222141861915588,
   0.5904055237770081,
   0.5552389025688171,
   0.7874670624732971,
   0.6510953307151794,
   0.4674784541130066,
   0.8590373396873474,
   0.7565814852714539,
   0.6861440539360046,
   0.7143733501434326,
   0.5823841691017151,
   0.8986912965774536,
   0.6324862241744995,
   0.7275481820106506,
   -1.3297263383865356,
   0.6372591853141785,
   0.48886215686798096,
   0.9801378846168518,
   0.7302512526512146,
   0.6641771197319031,
   0.7125130295753479,
   0.613391101360321,
   0.5709782838821411,
   0.8295596241950989,
   0.5886173844337463,
   0.5672508

не помогло, а если добавить параметр p?

In [51]:
k_source = []
k_hypothesis = []
k_reference = []
k_compare = {}

Длину всегда брали максимальную, чтобы не резать тексты, а если текст короче, то остановится автоматически.
топ к равно 50, то есть наш sampling pool органичен до 50 наиболее вероятных слов, из которых будем выбирать;
при этом мы еще ставим параметр top_p = 0.95, это вероятность, которая нас устроит
топ p и топ k на практике лучше совмещать, проверим, что получится, и будет ли лучше

In [53]:
for i in range(len(target)):

  target_item = target[i] # reference
  #print(target_item, 'TARGET')
  src_item = src[i] # source
  #print(src_item, 'src_item')
  inputs = tokenizer.encode(src_item, return_tensors="pt", max_length=512, truncation=True)
  #print(inputs, 'inputs')
  
  sample_output = model.generate(
    inputs.cuda(), 
    do_sample=True, 
    max_length=512, 
    top_k=50, 
    top_p=0.95, 
    num_return_sequences=3,
    early_stopping = True
)
  result = tokenizer.decode(sample_output[0], skip_special_tokens=True) # hypothesis
  k_compare[target_item] = result
  #print(result, 'RESULT')

  # collecting data for metrics count
  k_source.append(src_item)
  k_hypothesis.append(result)
  k_reference.append([target_item])
  

/usr/local/lib/python3.7/dist-packages/transformers/generation_utils.py:2138: UserWarning: __floordiv__ is deprecated, and its behavior will change in a future version of pytorch. It currently rounds toward 0 (like the 'trunc' function NOT 'floor'). This results in incorrect rounding for negative values. To keep the current behavior, use torch.div(a, b, rounding_mode='trunc'), or for actual floor division, use torch.div(a, b, rounding_mode='floor').
  next_indices = next_tokens // vocab_size


In [54]:
k_source

['The first in a series of simulated exercises will show the critical role rapid-response mobile laboratories, known as Public Health RRMLs, play in fighting public health emergencies by enabling onsite testing and analysis during outbreaks.',
 'Ten days after the implementation of broad public health and social measures, cases began to decline in Germany. Mortality rates and the median age of confirmed cases in Germany are lower than the average elsewhere. This is linked to a range of factors – including population demography and widespread testing.',
 'Mental health challenges remain deeply stigmatized in many communities. Community sensitization to the spectrum of mental health challenges will be important as the true scale of the crisis becomes clearer. Specialized care and support services need to be sustained for those in the grip of more severe mental health challenges.',
 'It is therefore vital to take a long-term view and to use that perspective to strengthen health policies a

In [55]:
k_hypothesis

['Die erste in einer Reihe von Simulationsübungen soll die wichtige Rolle verdeutlichen, die mobile Labore für den schnellen Feldeinsatz, sogenannte Rapid-Response Mobile Laboratories (RRML), bei der Eindämmung gesundheitlicher Notlagen spielen, indem sie während Krankheitsausbrüchen die Durchführung von Tests und Analysen direkt vor Ort ermöglichen.',
 'Zehn Tage nach Einführung breiter gesundheitlicher und sozialer Maßnahmen sank die Zahl der Fälle in Deutschland. Die Sterblichkeitsraten und das Durchschnittsalter der bestätigten Fälle sind in Deutschland niedriger als im Durchschnitt andererorts. Dies geht mit einer Reihe von Faktoren einher, darunter die demografische Demografie und die flächendeckende Testung.',
 'Psychische Herausforderungen bleiben in vielen Gemeinschaften zutiefst stigmatisiert. Die Sensibilisierung der Gemeinschaft für das Spektrum der psychischen Herausforderungen wird von Bedeutung sein, da die wahre Größenordnung der Krise klarer wird. Spezialisierte Versor

In [56]:
k_reference

[['Die erste in einer Reihe von Simulationsübungen soll die wichtige Rolle verdeutlichen, die mobile Labore für den schnellen Feldeinsatz, sogenannte Rapid-Response Mobile Laboratories (RRML), bei der Eindämmung gesundheitlicher Notlagen spielen, indem sie während Krankheitsausbrüchen die Durchführung von Tests und Analysen direkt vor Ort ermöglichen.'],
 ['Doch auch innerhalb von Ländern sind die Fälle nicht gleichmäßig verteilt. So tragen in Italien die Lombardei und in Spanien Madrid jeweils die Hauptlast der Pandemie. Glücklicherweise nimmt die Zahl der täglichen Todesfälle in beiden Regionen inzwischen ab.'],
 ['Psychische Gesundheitsprobleme sind in unseren Gesellschaften oftmals noch mit einem starken Stigma verbunden. Eine Sensibilisierung der Gesellschaft für die große Bandbreite an psychischen Gesundheitsproblemen wird nun, da sich das volle Ausmaß der Krise allmählich offenbart, immer wichtiger. Für Personen, die unter schwereren psychischen Problemen leiden, müssen speziali

In [30]:
get_results('k_p_samples_results.json', k_source,k_hypothesis,k_reference)

[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package omw-1.4 to /root/nltk_data...
[nltk_data]   Package omw-1.4 is already up-to-date!
GPU available: False, used: False
TPU available: False, using: 0 TPU cores
Predicting: 100%|██████████| 13/13 [00:18<00:00,  1.39s/it]


{'bleu': {'bleu': 0.53570232579626,
  'brevity_penalty': 0.8156713097207712,
  'length_ratio': 0.830741549929315,
  'precisions': [0.8031404702970297,
   0.6794512004989086,
   0.6098365807668134,
   0.5590750712701932],
  'reference_length': 15562,
  'translation_length': 12928},
 'comet': ([1.2384836673736572,
   0.38465139269828796,
   0.44683322310447693,
   0.4034285247325897,
   1.0452475547790527,
   0.1493382453918457,
   1.301583170890808,
   1.613797903060913,
   1.3787167072296143,
   -0.24043112993240356,
   0.41493797302246094,
   0.8764432668685913,
   0.7350091934204102,
   1.0508575439453125,
   -0.5646752119064331,
   0.30187976360321045,
   0.4321848750114441,
   0.8287811875343323,
   0.798665463924408,
   0.46910393238067627,
   0.8736383318901062,
   0.8564619421958923,
   0.10823822021484375,
   0.3345126211643219,
   0.7024510502815247,
   0.2560124099254608,
   0.29073503613471985,
   -0.8884382843971252,
   0.7105202078819275,
   0.6022520661354065,
   0.693115

bleu: 0.53570232579626

comet: 0.6077582398056984

meteor: 0.6265148442605895

sacrebleu: 46.5377863718655

о, полегчало